File: python/be/bot/client/bot.py
"""A market-maker over the mini-cex spot order book. Mirror of node/be/bot/client/bot.js.

Quotes a post-only bid below and ask above a reference price, cancels/replaces on
reprice, passes each order past the pure RiskEngine, retries only transport
failures, and never leaks its token. Prices are in the order book's integer form
(quote-per-base x 1e8).
"""

from __future__ import annotations

from be.api.venues.minicex import ApiUnreachable
from be.bot.risk.engine import RiskEngine

SCALE = 100_000_000


class MarketMaker:
    def __init__(self, cex, token=None, handle=None, base=None, quote=None, ref_price=None,
                 spread_bps=50, size=None, dry_run=False, kill_switch=False, risk=None):
        self.cex = cex
        self.token = token
        self.handle = handle
        self.base = base
        self.quote_asset = quote
        self.ref_price = ref_price
        self.spread_bps = spread_bps
        self.size = size
        self.dry_run = dry_run
        self.kill_switch = kill_switch
        self.risk = risk or RiskEngine(kill_switch=kill_switch)

    def bid_price(self):
        return (self.ref_price * (10000 - self.spread_bps)) // 10000

    def ask_price(self):
        return (self.ref_price * (10000 + self.spread_bps)) // 10000

    def risk_ctx(self):
        assets = {a["symbol"]: {"usd_micro": a["usd_micro"]} for a in self.cex.get("/assets")["body"]["assets"]}
        balances = {b["asset"]: b["amount"] for b in self.cex.get("/balances/" + self.handle)["body"]["balances"]}
        pairs = self.cex.get("/pairs")["body"]["pairs"]
        return {"assets": assets, "balances": balances, "pairs": pairs}

    def cancel_all(self):
        r = self.cex.orders_of(self.handle)
        for o in (r["body"]["orders"] or []):
            if o["base"] == self.base and o["quote"] == self.quote_asset and o["status"] in ("open", "partial"):
                self.cex.request("DELETE", "/orders/" + str(o["id"]), token=self.token)

    def active_quotes(self):
        r = self.cex.orders_of(self.handle, "open")
        return [o for o in (r["body"]["orders"] or []) if o["base"] == self.base and o["quote"] == self.quote_asset]

    def quote(self):
        if self.kill_switch:
            return {"ok": False, "reason": "kill switch"}
        bid, ask = self.bid_price(), self.ask_price()
        ctx = self.risk_ctx()
        p_buy = self.risk.plan({"kind": "order", "side": "buy", "base": self.base, "quote": self.quote_asset, "size": self.size, "price": bid}, ctx)
        if not p_buy["ok"]:
            return {"ok": False, "reason": p_buy["reason"], "leg": "bid"}
        p_sell = self.risk.plan({"kind": "order", "side": "sell", "base": self.base, "quote": self.quote_asset, "size": self.size, "price": ask}, ctx)
        if not p_sell["ok"]:
            return {"ok": False, "reason": p_sell["reason"], "leg": "ask"}
        if self.dry_run:
            return {"ok": True, "dryRun": True, "bid": bid, "ask": ask}
        self.cancel_all()
        bid_res = self.cex.place_order(self.token, {"base": self.base, "quote": self.quote_asset, "side": "buy", "type": "limit", "price": bid, "size": self.size, "post_only": True})
        ask_res = self.cex.place_order(self.token, {"base": self.base, "quote": self.quote_asset, "side": "sell", "type": "limit", "price": ask, "size": self.size, "post_only": True})
        return {"ok": True, "bid": bid_res["body"], "ask": ask_res["body"], "bidPrice": bid, "askPrice": ask}

    def reprice(self, new_ref):
        self.ref_price = new_ref
        return self.quote()

    def with_retry(self, fn, tries=3):
        last = None
        for _ in range(tries):
            try:
                return fn()
            except ApiUnreachable as err:
                last = err
        raise last

    def leaked_secret(self, s):
        return self.token in str(s)

File: node/be/bot/client/bot.js
'use strict';

const { ApiUnreachable } = require('../../api/venues/minicex');
const { RiskEngine } = require('../risk/engine');

/**
 * A market-maker that quotes a two-sided spread on one pair of the mini-cex spot
 * order book. It posts a post-only bid below and ask above a reference price,
 * cancels and replaces its quotes on a reprice, and passes every order past the
 * pure RiskEngine before sending. Mirrors the pattern of the perpetuals repo's
 * Bot client: it retries only transport failures, never a valid rejection, and
 * never lets its token leak.
 *
 * Prices are in the order book's integer form (quote base-units per whole base,
 * i.e. a human price x 1e8); spread math is BigInt so the reference x 10000 does
 * not overflow. The bot is otherwise pure orchestration over the venue.
 */
class MarketMaker {
  constructor(cex, opts = {}) {
    this.cex = cex;
    this.token = opts.token;
    this.handle = opts.handle;
    this.base = opts.base;
    this.quoteAsset = opts.quote;
    this.refPrice = opts.refPrice;               // integer, quote-per-base x 1e8
    this.spreadBps = opts.spreadBps === undefined ? 50 : opts.spreadBps;
    this.size = opts.size;                        // base units
    this.dryRun = !!opts.dryRun;
    this.killSwitch = !!opts.killSwitch;
    this.risk = opts.risk || new RiskEngine({ killSwitch: this.killSwitch });
  }

  bidPrice() { return Number((BigInt(this.refPrice) * BigInt(10000 - this.spreadBps)) / 10000n); }
  askPrice() { return Number((BigInt(this.refPrice) * BigInt(10000 + this.spreadBps)) / 10000n); }

  /** Build the risk context from the account's current balances + the venue's assets/pairs. */
  async riskCtx() {
    const assets = {}; for (const a of (await this.cex.get('/assets')).body.assets) assets[a.symbol] = { usd_micro: a.usd_micro };
    const balances = {}; for (const b of (await this.cex.get('/balances/' + this.handle)).body.balances) balances[b.asset] = b.amount;
    const pairs = (await this.cex.get('/pairs')).body.pairs;
    return { assets, balances, pairs };
  }

  async cancelAll() {
    const r = await this.cex.ordersOf(this.handle);
    for (const o of (r.body.orders || [])) if (o.base === this.base && o.quote === this.quoteAsset && (o.status === 'open' || o.status === 'partial')) await this.cex.cancelOrder(this.token, o.id);
  }
  async activeQuotes() {
    const r = await this.cex.ordersOf(this.handle, 'open');
    return (r.body.orders || []).filter((o) => o.base === this.base && o.quote === this.quoteAsset);
  }

  /**
   * Post (or refresh) the two-sided quote. Returns a decision: refused when the
   * kill switch is on or the risk gate blocks a leg; planned-only under dryRun;
   * otherwise the placed bid and ask.
   */
  async quote() {
    if (this.killSwitch) return { ok: false, reason: 'kill switch' };
    const bid = this.bidPrice(); const ask = this.askPrice();
    const ctx = await this.riskCtx();
    const pBuy = this.risk.plan({ kind: 'order', side: 'buy', base: this.base, quote: this.quoteAsset, size: this.size, price: bid }, ctx);
    if (!pBuy.ok) return { ok: false, reason: pBuy.reason, leg: 'bid' };
    const pSell = this.risk.plan({ kind: 'order', side: 'sell', base: this.base, quote: this.quoteAsset, size: this.size, price: ask }, ctx);
    if (!pSell.ok) return { ok: false, reason: pSell.reason, leg: 'ask' };
    if (this.dryRun) return { ok: true, dryRun: true, bid, ask };
    await this.cancelAll();
    const bidRes = await this.cex.placeOrder(this.token, { base: this.base, quote: this.quoteAsset, side: 'buy', type: 'limit', price: bid, size: this.size, post_only: true });
    const askRes = await this.cex.placeOrder(this.token, { base: this.base, quote: this.quoteAsset, side: 'sell', type: 'limit', price: ask, size: this.size, post_only: true });
    return { ok: true, bid: bidRes.body, ask: askRes.body, bidPrice: bid, askPrice: ask };
  }

  async reprice(newRef) { this.refPrice = newRef; return this.quote(); }

  /** Retry ONLY a transport failure (ApiUnreachable); a valid rejection is returned as-is. */
  async withRetry(fn, tries = 3) {
    let last;
    for (let i = 0; i < tries; i++) {
      try { return await fn(); }
      catch (err) { if (!(err instanceof ApiUnreachable)) throw err; last = err; }
    }
    throw last;
  }

  leakedSecret(s) { return String(s).includes(this.token); }
}

module.exports = { MarketMaker };
